### ISSUE TITLE
Recursive type bug

### ISSUE THREAD

**Issue by tamuhey** [microsoft/pyright#1102]
```python
from typing import Dict, Union, cast


A = Union[str, Dict[str, "A"]]


def foo(x: A):
    if isinstance(x, str):
        print(x)
    else:
        for _, v in x.items():
            # v's type is `A`, it seems to be correctly inferred
            foo(v)  # error
            foo(cast(A, v))  # ok
            v: A = v  # ok


```

```
  13:17 - error: Argument of type "A" cannot be assigned to parameter "x" of type "A" in function "foo"
    Type "A" cannot be assigned to type "A"
      Type "Type[str]" cannot be assigned to type "A"
        Type "Type[str]" cannot be assigned to type "str"
        Type "Type[str]" cannot be assigned to type "Dict[str, A]"
      Type "Type[Dict[str, A]]" cannot be assigned to type "A"
        Type "Type[Dict[str, A]]" cannot be assigned to type "str"
        Type "Type[Dict[str, A]]" cannot be assigned to type "Dict[str, A]" (reportGeneralTypeIssues)
```

`v` seems to be correctly inferred as type `A`.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.80, which I just published. It will also be in the next release of pylance.
